lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -864,18 +872,44 @@ class VerilogEmitterState {
   /// Whether a newline is expected, emitted late to provide opportunity to
   /// open/close boxes we don't know we need at level of individual statement.
   /// Every statement should set this instead of directly emitting (last)
-  /// newline. Most statements end with emitLocationInfoAndNewLine which handles
-  /// this.
+  /// newline. Most statements end with emitLocationInfoAndNewLine which
+  /// handles this.
   bool pendingNewline = false;
 
   /// String storage backing Tokens built from temporary strings.
   /// PrettyPrinter will clear this as appropriate.
-  TokenStringSaver saver;
+  TokenStringAndCallbackSaver saver;
 
   /// Pretty printer.
   PrettyPrinter pp;
 
+  // Name of the output file, used for debug information.
+  StringAttr fileName;
+
+  CallbackType *getCallback(Operation *op, OpLocMap *map,
+                            bool beginPrint = true) {
+    if (!op)
+      return nullptr;
+    llvm::formatted_raw_ostream *formattedStream = &os;
+    auto callback = [op, formattedStream, map, beginPrint]() {
+      map->addLoc(op, *formattedStream, beginPrint);
+    };
+    return saver.save(callback);
+  }
+  CallbackToken getCallbackToken(Operation *op, bool beginPrint = true) {
+    if (options.noVerilogLocation)
+      return CallbackToken(nullptr);
```

Comment:
Could `CallbackToken` never have a null callback? Seems weird to have to test for it, vs just never constructing (or feeding) it to the PP?